api/grpcserver/v2alpha1/activation.go

```diff
@@ -64,27 +65,50 @@ func (s *ActivationStreamService) Stream(
 			return status.Errorf(codes.Unavailable, "can't send header")
 		}
 	}
-	ops, err := toOperations(toRequest(request))
+
+	dbChan := make(chan *types.VerifiedActivationTx, 100)
+	errChan := make(chan error, 1)
+
+	ops, err := toAtxOperations(toAtxRequest(request))
 	if err != nil {
 		return status.Error(codes.InvalidArgument, err.Error())
 	}
-	var ierr error
-	if err := atxs.IterateAtxsOps(s.db, ops, func(atx *types.VerifiedActivationTx) bool {
-		ierr = stream.Send(&spacemeshv2alpha1.Activation{Versioned: &spacemeshv2alpha1.Activation_V1{V1: toAtx(atx)}})
-		return ierr == nil
-	}); err != nil {
-		return status.Error(codes.Internal, err.Error())
-	}
-	if sub == nil {
-		return nil
+
+	// send db data to chan to avoid buffer overflow
+	go func() {
+		defer close(dbChan)
+		if err := atxs.IterateAtxsOps(s.db, ops, func(atx *types.VerifiedActivationTx) bool {
+			select {
+			case dbChan <- atx:
+				return true
+			case <-ctx.Done():
+				// exit if the stream context is cancelled
+				return false
+			}
+		}); err != nil {
+			select {
+			case errChan <- status.Error(codes.Internal, err.Error()):
+			default:
+				ctxzap.Error(stream.Context(), "unable to send error", zap.Error(err))
+				select {
+				case <-ctx.Done(): // check ctx.Done() to avoid blocking
+				default:
+				}
```

Comment:
I think the `default` part can never happen because the channel is buffered, consider:
```suggestion
			errChan <- status.Error(codes.Internal, err.Error()):
```